### DISCUSSION TITLE
Omnet crashing on hasGcl = true

### DISCUSSION THREAD

**Discussion by paintspot87** [omnetpp/omnetpp#1118] [Q&A] [answered]
I'm trying to implement TSN time aware shaper with a GCL similarly to how I've done it on a real switch where I would define time-slots and what gates are open in that time-slot. Rather than defining durations per gate as in the showcase.

When I run the modified inet\showcases\tsn\trafficshaping\timeawareshaper showcase with `*.switch.eth[*].macLayer.queue.hasGcl = true` it just crashes and no matter what configuration I'm trying it just crashes.

I'm running omnet 6.0.2 with INET 4.5.2 on Windows 10.

Any ideas?

This is the .ini-file:

```
[General]
network = inet.networks.tsn.TsnLinearNetwork
sim-time-limit = 1s
description = "Traffic shaping using time-aware shapers"

**.displayGateSchedules = true
**.gateFilter = "**.eth[1].**"
**.gateScheduleVisualizer.height = 16
**.gateScheduleVisualizer.placementHint = "top"

# client applications
*.client.numApps = 2
*.client.app[*].typename = "UdpSourceApp"
*.client.app[0].display-name = "best effort"
*.client.app[1].display-name = "video"
*.client.app[*].io.destAddress = "server"
*.client.app[0].io.destPort = 1000
*.client.app[1].io.destPort = 1001
*.client.app[*].source.packetLength = 1000B - 54B # 54B = 8B (UDP) + 20B (IP) + 14B (ETH MAC) + 4B (ETH FCS) + 8B (ETH PHY)
*.client.app[0].source.productionInterval = exponential(200us) # ~40Mbps
*.client.app[1].source.productionInterval = exponential(400us) # ~20Mbps

# server applications
*.server.numApps = 2
*.server.app[*].typename = "UdpSinkApp"
*.server.app[0].io.localPort = 1000
*.server.app[1].io.localPort = 1001

# enable outgoing streams
*.client.hasOutgoingStreams = true

# client stream identification
*.client.bridging.streamIdentifier.identifier.mapping = [{stream: "best effort", packetFilter: expr(udp.destPort == 1000)},
                                                         {stream: "video", packetFilter: expr(udp.destPort == 1001)}]

# client stream encoding
*.client.bridging.streamCoder.encoder.mapping = [{stream: "best effort", pcp: 0},
                                                 {stream: "video", pcp: 4}]

# enable egress traffic shaping
*.switch.hasEgressTrafficShaping = true

# disable forwarding IEEE 802.1Q C-Tag
*.switch.bridging.directionReverser.reverser.excludeEncapsulationProtocols = ["ieee8021qctag"]

# time-aware traffic shaping
*.switch.eth[*].macLayer.queue.hasGcl = true
*.switch.eth[*].macLayer.queue.gateControlList.numGates = 8
*.switch.eth[*].macLayer.queue.gateControlList.durations = [4ms, 2ms]
*.switch.eth[*].macLayer.queue.gateControlList.gateStates = ["10000000", "01111111"]
#*.switch.eth[*].macLayer.queue.numTrafficClasses = 2
#*.switch.eth[*].macLayer.queue.*[0].display-name = "best effort"
#*.switch.eth[*].macLayer.queue.*[1].display-name = "video"
#*.switch.eth[*].macLayer.queue.transmissionGate[0].offset = 0ms
#*.switch.eth[*].macLayer.queue.transmissionGate[0].durations = [4ms, 6ms] # period is 10
#*.switch.eth[*].macLayer.queue.transmissionGate[1].offset = 6ms
#*.switch.eth[*].macLayer.queue.transmissionGate[1].durations = [2ms, 8ms]
```

**Comment by levy** [comment, marked as answer]
This issue is already fixed in the INET master branch.

**Reply by paintspot87** [reply]
I see! I will check it out. Thank you.

**Reply by Frt0214** [reply]
I also encountered a similar problem, how to solve it? Thank you so much!

**Reply by paintspot87** [reply]
@Frt0214 It was solved by using the master branch as levy said.

Personally I just replaced 

```
GateControlList.cc
GateControlList.h
GateControlList.ned
```

and it works now.

https://github.com/inet-framework/inet/tree/881fbe62eea258c3c8775659fb34cf511bfa6bda/src/inet/queueing/gate

**Reply by Frt0214** [reply]
@paintspot87  The mistake is no longer happening. Thank you for your help!

**Reply by Frt0214** [reply]
*.switch1.eth[3].macLayer.queue.numTrafficClasses = 3
*.switch1.eth[3].macLayer.queue.hasGcl=true
*.switch1.eth[3].macLayer.queue.gateControlList.numGates=3
*.switch1.eth[3].macLayer.queue.gateControlList.offsets = 8936us
*.switch1.eth[3].macLayer.queue.gateControlList.gateStates=["001","110","000"]
*.switch1.eth[3].macLayer.queue.gateControlList.durations = [8us,984us,8us]
*.switch1.eth[3].macLayer.queue.gateControlList.mapping="012"
.switch1.eth[3].macLayer.queue.[0].display-name = "interface"
.switch1.eth[3].macLayer.queue.[1].display-name = "be"
.switch1.eth[3].macLayer.queue.[2].display-name = "tt"

@levy @paintspot87 , hi ! Above is one of my gating configurations. When I set hasgcl to true, can I configure the offset parameter for the gcl entry? The offset parameter of the entire table appears to be missing. It appears that the offset parameter is invalid. In other words, how do I make the entire entry run after the offset value?

Best Regards

**Reply by Frt0214** [reply]
Just like basetime>currenttime indicates that the gate entry is running

**Reply by levy** [reply]
There's no offset parameter in GateControlList and you cannot set the offset parameter of PeriodicGate modules when using the GateControlList.
